v2transport/transport.go

```diff
@@ -0,0 +1,625 @@
+	// party sends.
+	garbageSize = 16
+
+	// maxGarbageLen is the maximum size of garbage that either peer is allowed
+	// to send.
+	maxGarbageLen = 4095
+
+	// maxContentLen is the maximum length of content that can be encrypted.
+	maxContentLength = 1<<24 - 1
+
+	// lengthFieldLen is the length of the length field when encrypting the
+	// content's length.
+	lengthFieldLen = 3
+
+	// headerLen is the length of the header field. It is composed of a single
+	// byte with only the ignoreBitPos having any meaning.
+	headerLen = 1
+
+	// chachapoly1305Expansion is the difference in bytes between the plaintext
+	// and ciphertext when using chachapoly1305. The ciphertext is larger
+	// because of the authentication tag.
+	chachapoly1305Expansion = 16
+)
+
+var (
+	// transportVersion is the transport version we are currently using.
+	transportVersion = []byte{}
+
+	// errInsufficientBytes is returned when we haven't received enough bytes
+	// to populate their ElligatorSwift encoded public key.
+	errInsufficientBytes = fmt.Errorf("insufficient bytes received")
+
+	// ErrUseV1Protocol is returned when the initiating peer is attempting to
+	// use the V1 protocol.
+	ErrUseV1Protocol = fmt.Errorf("use v1 protocol instead")
+
+	// errWrongNetV1Peer is returned when a v1 peer is using the wrong network.
+	errWrongNetV1Peer = fmt.Errorf("peer is v1 and using the wrong network")
+
+	// errGarbageTermNotRecv is returned when a v2 peer never sends us their
+	// garbage terminator.
+	errGarbageTermNotRecv = fmt.Errorf("no garbage term received")
+
+	// errContentLengthExceeded is returned when trying to encrypt more than
+	// the maximum content length.
+	errContentLengthExceeded = fmt.Errorf("maximum content length exceeded")
+
+	// errFailedToRecv is returned when a Read call fails.
+	errFailedToRecv = fmt.Errorf("failed to recv data")
+
+	// errPrefixTooLarge is returned if receivedPrefix is ever too large. This
+	// shouldn't happen unless the API is mis-used.
+	errPrefixTooLarge = fmt.Errorf("prefix too large - internal error")
+
+	// errGarbageTooLarge is returned if a caller attempts to send garbage
+	// larger than normal.
+	errGarbageTooLarge = fmt.Errorf("garbage too large")
+)
+
+// Peer defines the components necessary for sending/receiving data over the v2
+// transport.
+type Peer struct {
+	// privkeyOurs is our private key
+	privkeyOurs *btcec.PrivateKey
+
+	// ellswiftOurs is our ElligatorSwift-encoded public key.
+	ellswiftOurs [64]byte
+
+	// sentGarbage is the garbage sent after the public key. This may be up to
+	// 4095 bytes.
+	sentGarbage []byte
+
+	// receivedPrefix is used to determine which transport protocol we're
+	// using.
+	receivedPrefix []byte
+
+	// sendL is the cipher used to send encrypted packet lengths.
+	sendL *FSChaCha20
+
+	// sendP is the cipher used to send encrypted packets.
+	sendP *FSChaCha20Poly1305
+
+	// sendGarbageTerm is the garbage terminator that we send.
+	sendGarbageTerm []byte
```

Comment:
IIUC this is always 16 bytes? So can make this a fixed size array.

Reply:
Yup